ocean-client/src/vault-maxi.ts

```diff
@@ -76,34 +73,50 @@ export async function main(event: maxiEvent): Promise<Object> {
     }
 
     let vault: LoanVaultActive = vaultcheck
+    let result = true
 
     // 2022-03-08 Krysh: Something went wrong on last execution, we need to clean up, whatever was done
-    if (settings.state !== VaultMaxiState.Start && settings.state !== VaultMaxiState.Finish) {
-        console.log("something went wrong on last execution, we clean up whatever was done")
-        let result = await program.cleanUp(vault, telegram)
-        await store.updateToState(VaultMaxiState.Start)
-        const cleanUpVaultCheck = await program.getVault() as LoanVaultActive
-        await telegram.log("executed clean-up part of script " + (result ? "successfull" : "with problems") + ". vault ratio after clean-up " + cleanUpVaultCheck.collateralRatio)
-        return {
-            statusCode: result ? 200 : 500
+    if (settings.stateInformation && settings.stateInformation.state !== ProgramState.Waiting) {
+        const information = settings.stateInformation
+        console.log("last execution stopped state " + information.state)
+        console.log(" for tx " + information.tx)
+        console.log(" on block height " + information.blockHeight)
+
+        let shouldCleanUp = false
+
+        // 2022-03-09 Krysh: input of kuegi
+        // if we are on state waiting for last transaction and we are still on same block height since last execution
+        // then we should wait for next block
+        const currentBlockHeight = await program.getBlockHeight()
+        if (information.state === ProgramState.WaitingForLastTransaction && information.blockHeight === currentBlockHeight) {
+            console.log("waiting for next block")
+            await program.waitForBlockAfter(currentBlockHeight)
+        }
+        // 2022-03-09 Krysh: only clean up if it is really needed, otherwise we are fine and can proceed like normally
+        if (VaultMaxiProgram.shouldCleanUpBasedOn(information.tx as VaultMaxiProgramTransaction)) {
+            console.log("need to clean up")
+            result = await program.cleanUp(vault, telegram)
+            const cleanUpVaultCheck = await program.getVault() as LoanVaultActive
+            await telegram.log("executed clean-up part of script " + (result ? "successfull" : "with problems") + ". vault ratio after clean-up " + cleanUpVaultCheck.collateralRatio)
+
+            await program.updateToState(result ? ProgramState.Waiting : ProgramState.Error, VaultMaxiProgramTransaction.None)
```

Comment:
what would happen on case "State.Error, None" in the next execution of the script?

Reply:
within `program.updateToState` the current blockHeight will be saved too.
"error|none" will never happen, but yes it will just fall through, because my if is wrong.
good catch